### ISSUE TITLE
[minor] Race condition during build

### ISSUE THREAD

**Issue by tamlin-mike** [icsharpcode/ILSpy#967]
A minor problem. Restarting the build solves it. I just wanted to document it.

```
1>------ Build started: Project: Mono.Cecil, Configuration: net_4_0_Debug Any CPU ------
2>------ Build started: Project: ICSharpCode.TreeView, Configuration: Debug Any CPU ------
3>------ Build started: Project: ICSharpCode.Decompiler, Configuration: Debug Any CPU ------
4>------ Build started: Project: Mono.Cecil.Pdb, Configuration: net_4_0_Debug Any CPU ------

3>Out-File : The process cannot access the file '<snip>\ICSharpCode.Decompiler\ICSharpCode.Decompiler.nuspec' because it is being used by another process.
3>At <snip>\BuildTools\update-assemblyinfo.ps1:121 char:20
3>+             $out | Out-File -Encoding utf8 $file.Output;
3>+                    ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
3>    + CategoryInfo          : OpenError: (:) [Out-File], IOException
3>    + FullyQualifiedErrorId : FileOpenFailure,Microsoft.PowerShell.Commands.OutFileCommand
3>
3><snip>\ICSharpCode.Decompiler.csproj(590,5): error <snip>
```

**Comment by siegfriedpammer** [comment]
Seems this was introduced in 6faa5913c52d62016e1835aa8f68709016e08e9e ... is there another alternative to `$PSScriptRoot`?

**Comment by siegfriedpammer** [comment]
@tamlin-mike You can close this, if the problem has disappeared.

**Comment by tamlin-mike** [comment]
I think the problem is not the $PWD/$PSScriptRoot, but simply that some other part of the build system had the `.nuspec` file open for reading at the time the script ran. At least that's my understanding of the error message.

**Comment by siegfriedpammer** [comment]
We are using two target framework monikers in ICSharpCode.Decompiler.csproj: net461 and netstandard2.0

VS is executing the build for both frameworks at the same time. Because of this `update-assemblyinfo.ps1` is executed in parallel, i.e., two powershell processes tried to access the file.

The reason for me to suspect the recent change is that I had a similar error after I changed `$PSScriptRoot` to `Get-Location`. Weird stuff...

**Comment by tamlin-mike** [comment]
> Because of this update-assemblyinfo.ps1 is executed in parallel

Ah, that sounds like it might be it.

Anyway, it's nothing game-breaking for interactive builds, but since it could potentially make CI-builds fail intermittently I figured it best to flag for a potential problem.

**Comment by dgrunwald** [comment]
The problem was indeed `Get-Location`, which for some reason always returns a different (though identical-looking) value:

```
PS C:\work\ILSpy> (Get-Location).ToString()
C:\work\ILSpy
PS C:\work\ILSpy> $PWD.ToString()
C:\work\ILSpy
PS C:\work\ILSpy> (Get-Location).GetType().FullName
System.Management.Automation.PathInfo
PS C:\work\ILSpy> $PWD.GetType().FullName
System.Management.Automation.PathInfo
PS C:\work\ILSpy> (Get-Location).GetHashCode()
957762
PS C:\work\ILSpy> (Get-Location).GetHashCode()
62171659
PS C:\work\ILSpy> (Get-Location).GetHashCode()
44920701
PS C:\work\ILSpy> (Get-Location).GetHashCode()
2149042
PS C:\work\ILSpy> $PWD.GetHashCode()
52357250
PS C:\work\ILSpy> $PWD.GetHashCode()
52357250
PS C:\work\ILSpy> $PWD.GetHashCode()
52357250
```

**Comment by dgrunwald** [comment]
The race condition should be fixed in b78d640abdad1470a8175fa966ce0efd97651fb9.

**Comment by tamlin-mike** [comment]
I must say, this made me do my best impression of deer-in-headlights. The observable values of Get-Location *are* the same on consecutive invocations, even if you pipe them to `fl` there are no differences, so how they could have different hashes is... baffling.

Being a curious soul, I tried get to the bottom of this, and the fact that ($PWD).GetHashCode() from different command prompts returns different hashes seems to point in only one direction.

$PWD, being a built-in, is a process/instance singleton, while Get-Location (obviously) returns a new object every time. That combined with Microsoft failing to properly override GetHashCode(), results in this error. I followed the implementation type up to its very base type `InternalCommand`, and no GetHashCode() override was to be found anywhere, meaning the default in `object` is used.

Good sleuthing.
